### PR TITLE
Backport: Spark 12583

### PR DESCRIPTION
**Pull request by corruptmemory**
## This is a backport of the upstream fix for [SPARK-12583](https://github.com/apache/spark/pull/11272)

It's initially based on the patch from this commit: https://github.com/apache/spark/commit/310981d49a332bd329303f610b150bbe02cf5f87

After applying the commit (cherry-pick) I updated the source by selectively applying code from master.
## How was this patch tested?

Not tested yet.  Only builds.

### CODE REVIEW

**Comment by corruptmemory** [comment]
Rebased on latest in `branch-1.6` from upstream

**Comment by dragos** [comment]
It looks like to have Scala style check errors. You can run the exact same suite as the validator with `dev/run-tests`, or just the style-checks with `dev/lint-scala`

```
Scalastyle checks failed at following occurrences:
[error] /home/ubuntu/workspace/ghprb-spark-multi-conf/label/mesos-spark-docker/scala_version/2.11/core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:331:44: No space after token :
[error] (core/compile:scalastyle) errors exist
[error] Total time: 10 s, completed Apr 26, 2016 4:18:18 PM
[error] /home/ubuntu/workspace/ghprb-spark-multi-conf/label/mesos-spark-docker/scala_version/2.11/core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:331:44: No space after token :
[error] (core/compile:scalastyle) errors exist
[error] Total time: 10 s, completed Apr 26, 2016 4:18:50 PM
[error] running /home/ubuntu/workspace/ghprb-spark-multi-conf/label/mesos-spark-docker/scala_version/2.11/dev/lint-scala ; received return code 1
```

File: core/src/main/scala/org/apache/spark/rpc/RpcEndpointAddress.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * An address identifier for an RPC endpoint.
+ *
+ * The `rpcAddress` may be null, in which case the endpoint is registered via a client-only
+ * connection and can only be reached via the client that sent the endpoint reference.
+ *
+ * @param rpcAddress The socket address of the endpoint.
+ * @param name Name of the endpoint.
+ */
+private[spark] case class RpcEndpointAddress(val rpcAddress: RpcAddress, val name: String) {
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/rpc/RpcEndpointAddress.scala]: I think this whole file shouldn't be here. Note that there's been a lot of shuffling around this class. In 1.6.0 there are two RPC implementations: Netty-based and Akka-based. In 2.0 the Akka one is removed.

This class already exists in your fork under `org.apache.spark.rpc.netty`. You should probably modify the heartbeat logic to work with both RPC implementations and remove this copy. This commit is related: 4f5a24d

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala [hunk 2]
```diff
@@ -18,20 +18,20 @@
...
-import org.apache.mesos.Protos.{TaskInfo => MesosTaskInfo, _}
 import org.apache.mesos.{Scheduler => MScheduler, SchedulerDriver}
+import org.apache.mesos.Protos.{TaskInfo => MesosTaskInfo, _}
 
-import org.apache.spark.{SecurityManager, SparkContext, SparkEnv, SparkException, TaskState}
+import org.apache.spark.{SecurityManager, SparkContext, SparkException, TaskState}
 import org.apache.spark.network.netty.SparkTransportConf
 import org.apache.spark.network.shuffle.mesos.MesosExternalShuffleClient
-import org.apache.spark.rpc.RpcAddress
+import org.apache.spark.rpc.{RpcEndpointAddress}
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala]: Probably the same applies here...

**Comment by corruptmemory** [comment]
retest this please

**Comment by corruptmemory** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by corruptmemory** [comment]
Closing to send upstream

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by skonto** [comment]
retest this please

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**